teps/0121-unified-retries-for-runtime-objects.md

```diff
@@ -0,0 +1,421 @@
+- which performance characteristics that must be met?
+- which specific edge cases that must be handled?
+- which user scenarios that will be affected and must be accommodated?
+-->
+
+## Related Work
+
+In the CI/CD industry, most systems support retry, some of them enables users to configure the retry strategy[^retry-strategy].
+
+Typically, a retry strategy needs to specify:
+1. When to retry
+2. The amount of attempts
+3. Actions to take after a failed attempt
+4. Timeout of each attempt
+
+
+|  | [Retry Step in GA](https://github.com/marketplace/actions/retry-step) | [Retry Action in GA](https://github.com/marketplace/actions/retry-action) | [Argo Workflow](https://argoproj.github.io/argo-workflows/retries/#retry-policies) | [GitLab](https://docs.gitlab.com/ee/ci/yaml/#retry) | [Ansible](https://docs.ansible.com/ansible/latest/user_guide/playbooks_loops.html#retrying-a-task-until-a-condition-is-met)| [Concourse](https://concourse-ci.org/attempts-step.html#attempts-step) | Bamboo | Dagger |
+|:---|:---|:---|:---|:---|:---|:---|:---|:---|
+| **When to Retry** | configurable | on failure |configurable|configurable|[always retry, conditional stop](https://github.com/ansible/ansible/pull/76101) [^ansible-conditional-stop]|configurable|-|-|
+| **Attempts amount** |supported|supported|supported|supported|supported|supported|-|-|
+| **Timeout for each attempt** |supported|supported|supported|[supported](https://docs.gitlab.com/ee/ci/yaml/#retrywhen)|supported|supported|-|-|
+| **Timeout for all attempts** |[supported](https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#jobsjob_idstepstimeout-minutes)|[supported](https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#jobsjob_idstepstimeout-minutes)|[?](https://stackoverflow.com/questions/71424429/argo-workflowtemplate-timeout)|-|-|-|-|-|
+
+
+Several observations regarding of the feature table above:
+- We can configure **timeout duration per attempt** in all CI systems that support the `retry` functionality.
+- GitHub Action doesn't support retry natively, but because the flexibility of **customized actions**, some users write their own `retry` action to make it work, and those customized actions even support what to do before retrying a failed attempt.
+- Concourse mentioned the reason it [retries per attempt is somewhat arbitrary](https://concourse-ci.org/attempts-step.html#attempts-step).
+- Argo Workflow supports timeout per attempt, but [this stackoverflow question](https://stackoverflow.com/questions/71424429/argo-workflowtemplate-timeout) implies the need of "timeout for all retries".
+
+## Proposal
+
+This TEP proposes to unify the `retry` behavior as following:
+* Retry on all failures except for the cancellation.
+* The timeout is for each attempt.
+* It is the controller of each runtime object to implement retry functionality. (changeable, depending on how we deal with the large embedded retry status)
```

Comment:
hoping the work in improving embedded status from child runs in TEP-0100 can help here

https://github.com/tektoncd/community/blob/main/teps/0100-embedded-taskruns-and-runs-status-in-pipelineruns.md

Reply:
Thanks~ Put it under the references, we'll probably need to deal with the large CRD issue in the future.